Archipelago client (WIP)

Satisha10 opened the pull request:
Additions:
- Handle most of the packets (e.g. items received, items sent, data packages, messages).
- Detects when a location is checked, and when the game is finished.

TODO:
- Testing and debug.
- Parse the seed file and get the connection informations.

timoschwarzer commented on projects/Randomizer/archipelago/archipelago.cpp:
```diff
@@ -18,6 +28,39 @@ namespace randomizer::archipelago {
+            [this](const core::api::uber_states::UberStateCallbackParams& params, auto) {
+                if (params.state.get<bool>()) {
+                    // Game completion
+                    send_message(messages::StatusUpdate{30});
```
Where does the 30 come from? :eyes:

Satisha10 replied:
It is the value used by AP to indicate that the goal is completed (cf `ClientStatus` in the network protocol doc).

Satisha10 replied:
I added an enum structure in `archipelago_protocol.h` for the client status to make it more clear.

timoschwarzer commented on projects/Randomizer/archipelago/archipelago.h:
```diff
@@ -25,14 +25,426 @@ namespace randomizer::archipelago {
+        core::api::uber_states::UberState(14019, 20601),
+        core::api::uber_states::UberState(7228, 8370),
+        core::api::uber_states::UberState(10289, 44555),
+        core::api::uber_states::UberState(10289, 22102),
+        core::api::uber_states::UberState(14019, 26318),
+        core::api::uber_states::UberState(36153, 3013),
+        core::api::uber_states::UberState(0, 8),
+        core::api::uber_states::UberState(36153, 36521),
+        core::api::uber_states::UberState(36153, 12077),
+        core::api::uber_states::UberState(16155, 38979),
+        core::api::uber_states::UberState(16155, 49381),
+        core::api::uber_states::UberState(16155, 46270),
+        core::api::uber_states::UberState(48248, 4045),
+        core::api::uber_states::UberState(16155, 9230),
+        core::api::uber_states::UberState(16155, 55446),
+        core::api::uber_states::UberState(36153, 23902),
+        core::api::uber_states::UberState(36153, 3662),
+        core::api::uber_states::UberState(2, 1),
+        core::api::uber_states::UberState(2, 2),
+        core::api::uber_states::UberState(2, 3),
+        core::api::uber_states::UberState(2, 5),
+        core::api::uber_states::UberState(2, 19),
+        core::api::uber_states::UberState(2, 22),
+        core::api::uber_states::UberState(2, 26),
+        core::api::uber_states::UberState(2, 40),
+        core::api::uber_states::UberState(1, 23),
+        core::api::uber_states::UberState(1, 74),
+        core::api::uber_states::UberState(1, 98),
+        core::api::uber_states::UberState(1, 105),
+        core::api::uber_states::UberState(1, 106),
+        core::api::uber_states::UberState(1, 115),
+        core::api::uber_states::UberState(1, 116),
+        core::api::uber_states::UberState(1, 1074),
+        core::api::uber_states::UberState(1, 1098),
+        core::api::uber_states::UberState(1, 1106),
+        core::api::uber_states::UberState(1, 1115),
+        core::api::uber_states::UberState(1, 1116),
+        core::api::uber_states::UberState(15, 19396),
+        core::api::uber_states::UberState(15, 57987),
+        core::api::uber_states::UberState(15, 41666)        
```
Can we generate this at runtime from loc_data.csv instead of hardcoding it here?

Satisha10 replied:
It should be possible, I will look at that later.

Satisha10 replied:
It is now built in `archipelago.cpp` (I adapted your code from `archipelago_ids.cpp` for that).

timoschwarzer commented on projects/Randomizer/archipelago/archipelago.cpp:
```diff
@@ -18,6 +28,39 @@ namespace randomizer::archipelago {
     ArchipelagoClient::ArchipelagoClient() {
         m_websocket.setOnMessageCallback([this](const auto& msg) { on_websocket_message(msg); });
+        m_data_package_cache = read_file("ap_data_package.json");
+        m_item_id_to_name = read_file("ap_location_id_to_name.json");
+        m_location_id_to_name = read_file("ap_location_id_to_name.json");
```
Is this supposed to be reading the same file?

Satisha10 replied:
That was a typo, fixed.

timoschwarzer commented on projects/Randomizer/archipelago/archipelago.cpp:
```diff
@@ -99,20 +142,245 @@ namespace randomizer::archipelago {
+            [](const ids::ResourceItem& item) {
+                switch (item.type) {
+                    case ids::ResourceType::SpiritLight: {
+                        const auto assigner = std::make_shared<randomizer::seed::items::ValueModifier<int, randomizer::seed::items::ValueOperator::Add>>();
+                        assigner->variable = core::api::game::player::spirit_light();
+                        assigner->value.set(item.value);
+                        const auto collected = std::make_shared<randomizer::seed::items::ValueModifier<int, randomizer::seed::items::ValueOperator::Add>>();
+                        collected->variable = core::Property<int>(UberStateGroup::RandoStats, 3);
+                        collected->value.set(item.value);
```
```suggestion
                        const auto spirit_light = core::api::game::player::spirit_light();
                        const auto spirit_light_collected = core::api::uber_states::UberState(UberStateGroup::RandoStats, 3);
                        
                        spirit_light.set(spirit_light.get() + 1);
                        spirit_light_collected.set<int>(spirit_light_collected.get<int>() + 1);
```

Satisha10 replied:
Thank you ! I copied some code from `legacy_parser/parser.cpp` but misunderstood it. I also fixed the other resources.

timoschwarzer commented on projects/Randomizer/archipelago/archipelago.cpp:
```diff
@@ -4,6 +4,16 @@
 #include <Randomizer/archipelago/archipelago.h>
 #include <Randomizer/archipelago/archipelago_protocol.h>
 #include <Randomizer/archipelago/archipelago_save_meta.h>
+#include "Core/api/uber_states/uber_state.h"
```
```suggestion
#include <Core/api/uber_states/uber_state.h>
```

everywhere 🙂

Satisha10 replied:
Should be fixed (if I did not miss one of them).

timoschwarzer commented on projects/Randomizer/archipelago/archipelago.cpp:
```diff
@@ -99,20 +142,245 @@ namespace randomizer::archipelago {
+        modloader::info("archipelago", std::format("Write {} in ./archipelago.", file_name));
+    }
+
+    nlohmann::json ArchipelagoClient::read_file(const std::string& file_name) {
+        // TODO test this
+        std::ifstream i(m_data_package_path + file_name);
+        nlohmann::json j;
+        i >> j;
+        modloader::info("archipelago", std::format("Read data from {}.", file_name));
+        return j;
+    }
```
For reading files, we already have `read_text_file(...)` in `ext.h`. For paths, `std::filesystem::path` should be used instead of `std::string`.

timoschwarzer replied:
...we also have `load_json_file(...)` that takes a file name directly and handles prepending the base path and handling errors.

```cpp
nlohmann::json j;
if (!load_json_file("something.json", j)) {
  // Something went wrong
  // Error message is printed to modloader log
  return;
}
```

Satisha10 replied:
Thanks, I changed it to use `load_json_file`.

timoschwarzer commented on projects/Randomizer/archipelago/archipelago_protocol.cpp:
```diff
@@ -22,6 +23,26 @@ namespace randomizer::archipelago::messages {
         }
 
+        if (command == "LocationInfo") {
+        return message.get<LocationInfo>();
```
```suggestion
            return message.get<LocationInfo>();
```

💨

Satisha10 replied:
Fixed.

timoschwarzer commented on projects/Randomizer/archipelago/archipelago.cpp:
```diff
@@ -99,20 +142,245 @@ namespace randomizer::archipelago {
+        });
+    }
+
+    void ArchipelagoClient::write_file(const nlohmann::json& data, const std::string& file_name) {
+        // TODO verify that this works, and maybe save it somewhere else
+        std::ofstream o(m_data_package_path + file_name);
+        o << std::setw(4) << data << std::endl;
+        modloader::info("archipelago", std::format("Write {} in ./archipelago.", file_name));
+    }
```
Use `std::filesystem::path` and prepend `base_path()`.

Satisha10 replied:
Changed it.

timoschwarzer reviewed (changes requested):
Added some comments and TODOs here and there. 🙂

One thing that I noticed is that you never check against the `received_items` vector - you only push received items into it. I don't remember how this was supposed to be done right now, so this is just a reminder to remember this.

MunWolf commented on projects/Randomizer/archipelago/archipelago.cpp:
```diff
@@ -99,20 +142,245 @@ namespace randomizer::archipelago {
+        std::string game = m_slots[player.slot].game;
+        if (m_item_id_to_name[game].contains(item.item)) {  // TODO Maybe not necessary to check that
+            return m_item_id_to_name[game][item.item];
+            }
```
Should run clang-format on the files so the codestyle is correct

Satisha10 replied:
Thanks, I applied it to all the files.

MunWolf commented on projects/Randomizer/archipelago/archipelago.cpp:
```diff
@@ -99,20 +142,245 @@ namespace randomizer::archipelago {
+
+    std::string ArchipelagoClient::get_location_name(ids::archipelago_id_t id, const std::string& game) {
+        if (m_location_id_to_name[game].contains(id)) {  // TODO Maybe not necessary to check that
+            return m_location_id_to_name[game][id];
```
better to use m_location_id_to_name.find() here, then you aren't searching through it twice.

Satisha10 replied:
Thanks, I changed it in other places too.

MunWolf commented on projects/Randomizer/archipelago/archipelago_protocol.h:
```diff
@@ -3,9 +3,12 @@
+#include "Randomizer/archipelago/archipelago_ids.h"
 
 namespace randomizer::archipelago::messages {
+    // Data structures
```
Unnecessary comment

Satisha10 replied:
I removed it.

MunWolf commented on projects/Randomizer/archipelago/archipelago.cpp:
```diff
@@ -99,20 +142,245 @@ namespace randomizer::archipelago {
+                        adder->variable = core::api::game::player::max_health();
+                        adder->value.set(5);
+                        const auto refill = std::make_shared<randomizer::seed::items::Refill>();
+                        refill->type = randomizer::seed::items::Refill::RefillType::Health;
```
Unless I am missing something you are making these and then not doing anything with them?
Feel like there is a misunderstanding how the items work.

Satisha10 replied:
I copied some code from `legacy_parser/parser.cpp` without fully understanding it. I hopefully fixed that for all resources based on zre's suggestion.

MunWolf commented:
If I remember correctly the plan was to map archipelago location ids to uberstates and map our own locations to archipelago location ids.
And then just have a normal seed file take care of translating those into items (in which case you shouldn't need to work with randomizer::seed:items at all)
Might be wrong though, been a very long time since I was last involved in rando discussions, not to mention archipelago

Satisha10 commented:
> If I remember correctly the plan was to map archipelago location ids to uberstates and map our own locations to archipelago location ids. And then just have a normal seed file take care of translating those into items (in which case you shouldn't need to work with randomizer::seed:items at all) Might be wrong though, been a very long time since I was last involved in rando discussions, not to mention archipelago

The seed file only contains headers to setup the game file according to the options (to include `seed_core`, and for things like random spawn, no quests, goal mode, QOL). It does not contain any information about which item is in each pickup, this info is sent by the AP server after a pickup is collected (with a `LocationInfo` packet).

Satisha10 commented:
> One thing that I noticed is that you never check against the `received_items` vector - you only push received items into it. I don't remember how this was supposed to be done right now, so this is just a reminder to remember this.
I think that storing the items is not needed at all, since the only thing used to check for desyncs is the last item index (which increases by 1 for each received item), so I removed `received_items` and added the `last_item_index` in `archipelago_save_meta.cpp`.

timoschwarzer commented on projects/Randomizer/archipelago/archipelago_protocol.h:
```diff
@@ -184,7 +184,7 @@ namespace randomizer::archipelago::messages {
     struct InvalidPacket {
         std::string type;
-        std::string original_cmd;
+        std::string original_cmd;  // TODO can be null, which causes the parsing to fail
```
I *think* you can use `std::optional<std::string>` then to allow `null` values.

Satisha10 replied:
Apparently the `NLOHMANN_DEFINE_TYPE_INTRUSIVE` macro does not support conversion of optional, so I don't think it is an easy fix. I will just leave it like that for now since it is only a debug packet, and the full packet is still written `modloader_log` if the parsing fails.

timoschwarzer commented:
Will close this as it is merged by now. Thank you again!!
